client/lib/url/url-parts.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import { URL as URLString } from 'types';
+
+/**
+ * Internal dependencies
+ */
+import { determineUrlType, URL_TYPE } from './url-type';
+
+const BASE_URL = `http://__domain__.invalid`;
+
+interface UrlParts {
+	protocol: string;
+	host: string;
+	hostname: string;
+	port: string;
+	origin: string;
+	pathname: string;
+	hash: string;
+	search: string;
+	searchParams: URLSearchParams;
+	username: string;
+	password: string;
+}
+
+type UrlPartKey = keyof UrlParts;
+
+const EMPTY_URL: Readonly< UrlParts > = Object.freeze( {
```

Comment:
I'm not sure that it's our custom to use ALL_CAPS for constants. I know some people do, but personally I find them annoying in the same way that ALL CAPS WRITING IS ANNOYING. This isn't a simple defined constant either, which I think tends to see the APP_CAPS more often; it's an actual object.

There's no need to change this or `URL_PART_KEYS` but I thought I'd share my own thoughts on the matter. Not only are they actually `const` and so un-re-bindable, but w also have TypeScript to warn again any attempt to do so or to change their members. It's not like the languages of old whose type systems were so lacking that we had to write faux type systems via variable names that we manually check by inspection.

Reply:
Old habits die hard, I suppose. I'll leave this as-is for now, to avoid adding more changes, but may revisit this at some point 👍